File: 19.remove-nth-node-from-end-of-list.java
/*
 * @lc app=leetcode id=19 lang=java
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        return Solution1(head, n);
    }

    public ListNode Solution1(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        int size = 0;

        while (head != null) {
            size++;
            head = head.next;
        }

        head = dummy.next;
        int tailIndex = size - n;
        for (int i = 1; i < tailIndex; i++) {
            head = head.next;
        }

        // edge cases handler
        // remove the first element when size is equal to nth
        if (size == n) {
            dummy.next = dummy.next.next;
        } else {
            head.next = head.next.next;
        }

        return dummy.next;
    }

    public ListNode Solution2(ListNode head, int n) {
        ListNode fast, slow = head;
    }
}
// @lc code=end

File: 19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head) {
            return head;
        }

        ListNode* dummy = new ListNode(-1, head);
        ListNode* fast = dummy;
        ListNode* slow = dummy;

        for (int i = 0; fast; i++) {
            fast = fast->next;
            if (i > n) {
                slow = slow->next;
            }
        }

        slow->next = slow->next->next;

        return dummy->next;
    }
};
// @lc code=end

